File: src/python/PyAudioProcessor.py
from juce.juce_audio_processors import AudioProcessor
from juce.utils import AudioBuffer
from pprint import pprint


class PyAudioProcessor(AudioProcessor):
    pass
    # Todo: implement a simple plugin
    def prepareToPlay(self, sampleRate, samplesPerBlock):
        pass

    def releaseResources(self):
        pass

    def processBlock(self, buffer: AudioBuffer, midiMessages):
        buffer.applyGain(14)

    def createEditor(self):
        return None

    def hasEditor(self):
        return False

    def getName(self):
        return "PyAudioProcessor"

    def acceptsMidi(self):
        return True

    def producesMidi(self):
        return True

    def getTailLengthSeconds(self):
        return 0

    def getNumPrograms(self):
        return 0

    def getCurrentProgram(self):
        return 0

    def setCurrentProgram(self, index):
        pass

    def getProgramName(self, index):
        return ""

    def changeProgramName(self, index, newName):
        pass

    def getStateInformation(self, memoryBlock):
        pass

    def setStateInformation(self, data, sizeInBytes):
        pass



if __name__ == "__main__":
    proc = PyAudioProcessor()
    buff = AudioBuffer()


    # for method_name in dir(proc):
    #     method = getattr(proc, method_name)
    #     if callable(method) and not method_name.startswith("__"):
    #         for i in range(5):
    #             try:
    #                 method(*range(i))
    #                 break
    #             except TypeError:
    #                 continue
    #             except RuntimeError as e:
    #                 raise NotImplementedError("Method %s not implemented" % method_name) from None

File: venom/python/AudioProcessor/PyAudioProcessor.cpp
#include "include/PyAudioProcessor.h"

#include <thread>
#include <utility>

#include "spdlog/spdlog.h"

PyAudioProcessor::PyAudioProcessor(std::unique_ptr<py::object> clsInstance)
        : juce::AudioProcessor(),
          instance(std::move(clsInstance)) {
}

void PyAudioProcessor::prepareToPlay(double sampleRate, int samplesPerBlock) {
    spdlog::debug("[Prepare To Play]");

    py::gil_scoped_acquire acquire;
    instance->attr("prepareToPlay")(sampleRate, samplesPerBlock);
}

void PyAudioProcessor::releaseResources() {
    spdlog::debug("[Release Resources]");

    py::gil_scoped_acquire acquire;
    instance->attr("releaseResources")();
}

// TODO port BusesLayout
bool PyAudioProcessor::isBusesLayoutSupported(const BusesLayout &layouts) const {
    spdlog::debug("[BUS SUPPORT]");
    return false;
}

void PyAudioProcessor::processBlock(juce::AudioBuffer<float> &buffer, juce::MidiBuffer &midiMessages) {
    py::gil_scoped_acquire acquire;
    instance->attr("processBlock")(&buffer, &midiMessages);
}

juce::AudioProcessorEditor *PyAudioProcessor::createEditor() {
    JUCE_ASSERT_MESSAGE_MANAGER_IS_LOCKED

    py::gil_scoped_acquire acquire;
    auto editor = std::make_unique<py::object>(instance->attr("createEditor")());
    if (editor->is_none()) {
        return nullptr;
    }
    return new PyAudioProcessorEditor(this, std::move(editor));
}

bool PyAudioProcessor::hasEditor() const {
    py::gil_scoped_acquire acquire;
    return instance->attr("hasEditor")().cast<bool>();
}

const juce::String PyAudioProcessor::getName() const {
    py::gil_scoped_acquire acquire;
    return instance->attr("getName")().cast<juce::String>();
}

bool PyAudioProcessor::acceptsMidi() const {
    py::gil_scoped_acquire acquire;
    return instance->attr("acceptsMidi")().cast<bool>();
}

bool PyAudioProcessor::producesMidi() const {
    py::gil_scoped_acquire acquire;
    return instance->attr("producesMidi")().cast<bool>();
}

bool PyAudioProcessor::isMidiEffect() const {
    py::gil_scoped_acquire acquire;
    return instance->attr("isMidiEffect")().cast<bool>();
}

double PyAudioProcessor::getTailLengthSeconds() const {
    py::gil_scoped_acquire acquire;
    return instance->attr("getTailLengthSeconds")().cast<double>();
}

int PyAudioProcessor::getNumPrograms() {
    py::gil_scoped_acquire acquire;
    return instance->attr("getNumPrograms")().cast<int>();
}

int PyAudioProcessor::getCurrentProgram() {
    py::gil_scoped_acquire acquire;
    return instance->attr("getCurrentProgram")().cast<int>();
}

void PyAudioProcessor::setCurrentProgram(int index) {
    py::gil_scoped_acquire acquire;
    instance->attr("setCurrentProgram")(index);
}

const juce::String PyAudioProcessor::getProgramName(int index) {
    py::gil_scoped_acquire acquire;
    return instance->attr("getProgramName")(index).cast<juce::String>();
}

void PyAudioProcessor::changeProgramName(int index, const juce::String &newName) {
    py::gil_scoped_acquire acquire;
    instance->attr("changeProgramName")(index, newName);
}

void PyAudioProcessor::getStateInformation(juce::MemoryBlock &destData) {
    py::gil_scoped_acquire acquire;
    instance->attr("getStateInformation")(&destData);
}

void PyAudioProcessor::setStateInformation(const void *data, int sizeInBytes) {
    py::gil_scoped_acquire acquire;
    instance->attr("setStateInformation")(data, sizeInBytes);
}

PyAudioProcessor::~PyAudioProcessor() {
    spdlog::debug("[DESCTRUCTOR]");
    py::gil_scoped_acquire acquire;
    instance.reset();
}
